library/init/meta/interactive.lean

```diff
@@ -6,8 +6,11 @@ Authors: Leonardo de Moura
 prelude
 import init.meta.tactic init.meta.rewrite_tactic init.meta.simp_tactic
 import init.meta.smt.congruence_closure init.category.combinators
+import init.meta.lean_parser
 
 namespace interactive
+@[reducible] meta def parse {α : Type} (p : lean_parser α) : Type := α
```

Comment:
> Would you change all the primitive parsers and combinators to parser pexpr, or only the top-most parser?

All (primitive and user-defined) parsers would return `pexpr` (or `pexpr_for` if we implement that).

Okay, now that I read your proposal again, I completely missed the main trick: there are no open terms:
```lean
meta constant qexpr (rbp : num := std.prec.max) : lean_parser pexpr
```
Even if the we parse something with an open term, we always get back a pre-expression--and not an evaluated vm_obj.  Then we only need to convert back the lists and numbers and tuples, but not the pre-expressions.

If I understand you correctly, we would have the following definitions:
```lean
meta class has_quote (α : Type) := (quote : α -> pexpr)

meta instance (α β) [has_quote α] [has_quote β] : has_quote (α × β) :=
⟨ λ ⟨a, b⟩, mk_tuple_pexpr a b ⟩

meta instance (α) [has_quote α] : has_quote (list α) :=
⟨ mk_list_pexpr ⟩

-- Slight type change:
@[reducible] meta def parse {α : Type} [has_quote α] (p : lean_parser α) : Type := α
-- The intended semantics is that the parser first executes p,
-- then uses the has_quote instance to obtain a pexpr,
-- and then inserts the pexpr back in the AST.

-- And now the completely wrong instance for pexpr:
meta instance : has_quote pexpr := ⟨ id ⟩
```
I'm not sure I like that instance in a general quotation mechanism.  I might be okay if we only have that instance for `def antiquote_pexpr := pexpr`.

BTW, I think I found another problem with open terms (I'm completely fine with ignoring it):
```lean
meta def weird_parser (n : ℕ) : lean_parser bool :=
if n < 10 then tk "]" >> return tt else return ff

meta def weird_tactic (n : ℕ) : parse (weird_parser n) bool → tactic unit :=
sorry

meta example := λ x : ℕ, `[weird_tactic x]
-- what value of x does weird_parser see?
```